### DISCUSSION TITLE
Manual sync and diff for ApplicationSet applications

### DISCUSSION THREAD

**Discussion by Agalin** [argoproj/argo-cd#19358] [Ideas]
ApplicationSet is a really useful tool when it comes to deployments to multiple clusters but I find its sync configuration strangely limited.

If for one reason or another (testing, debugging, Kubernetes cluster burning down, whatever) I need to manually modify one of the managed applications I have those options:
* set sync policy to `create-only` or `create-delete`,
* set `ignoreApplicationDifferences`,
* set `preservedFields` (only for annotations and labels if I understand correctly),
* set whole applicationset controller to dry run.

All those options (except the last one) can be set on per-applicationset basis (assuming policy override is enabled). So far so good. I can configure what's needed then modify the underlying Application object.

So what's missing? Three things really.
* option to view how is specific Application differing from the ApplicationSet (equivalent of the diff view in the Application itself),
* option to list differing Applications (if you set any of those options there is no way to tell which Application has drifted),
* option to re-sync.

Last point is especially interesting because... it's technically doable. With sync policy set to `create-only` or `create-delete` you can delete the underlying Application object and controller will recreate it with clean state for you. That's not really a clean solution though.

I know, you might just re-enable automatic updates or remove ignores. Unfortunately that means you need to manually edit the ApplicationSet each time some kind of emergency change to the underlying Application happens. Is this situation rare? Probably, it definitely *should* be rare. It doesn't make it any less inconvenient, especially when you consider that UI doesn't show you why the field you've just modified has been automatically reverted (there is no automatic link in UI that would show which set manages the app, it only goes the other way around, from set's view to app's).

Yes, you can achieve that using App of Apps instead of ApplicationSet. But that means you lose access to a powerful templating machinery when it comes to cluster selection.

So, what do you think, is it feasible to extend ApplicationSet -> Application view in such way? Or maybe I'm missing some existing solutions after all?

**Comment by ChristianCiach** [comment]
It's kinda unfortunate that the ApplicationSet controller directly deploys the generated Application manifests to the cluster, isn't it? If there was a way to push the generated Application manifests to Git instead, we could just use an app-of-apps to solve pretty much every issue you mentioned.

I wonder if the upcoming "hydrator" feature could be used as a starting point to achieve something like that (because this will be the first time ArgoCD would natively be able to push stuff to Git).

**Reply by ChristianCiach** [reply]
I have an idea!

An upcoming version of the ArgoCD CLI will be able to generate a preview (dry-run) of the generated Application manifests:

* https://github.com/argoproj/argo-cd/pull/16781

So instead of deploying the ApplicationSet to the cluster, we could render the Application manifests outside the cluster, maybe as part of a CI pipeline and then push the rendered manifests back to Git. We could then use a regular app-of-apps to deploy the Applications to the cluster.

Obviously, this means that your CI system would need to connect to the ArgoCD server.

**Reply by Agalin** [reply]
I think that it could be solved if we go further: (a repo of) ApplicationSets usable as an Application source, just like Helm or Kustomize. Argo would basically need to run the preview generation logic under the hood to render manifests and pass them to the standard flow.

That may even be doable in a plugin with this CLI option. 🤔

**Reply by ChristianCiach** [reply]
You are right! This should also be doable as a ContentManagementPlugin (as soon as the CLI option becomes available).

This is surely not what the `--dry-run` option was intended for, but it really opens up some exciting new options.

**Comment by marxus** [comment]
@Agalin @ChristianCiach   https://github.com/marxus/argocd-appset
Use appset as cmp. all the benefits of app of apps and all the benefits of appset generators !

**Reply by ChristianCiach** [reply]
Awesome! That's pretty much what I had in mind when writing https://github.com/argoproj/argo-cd/discussions/19358#discussioncomment-10226690. Crazy that this actually works. The more I think about it, the more I believe that this a superior to the appset controller in every way.

**Reply by Agalin** [reply]
I didn't have time to experiment with this approach yet but with usage of Argo growing in our setup we'll sooner or later have to handle this case somehow. Will definitely check the project at that time. In the meantime we're switching branch on the whole appset which will cause an issue sooner or later.

I still believe native support would be the best way forward but still the plugin might be good enough. 🙂

**Reply by ChristianCiach** [reply]
> That's pretty much what I had in mind

Well, kinda. I didn't had an inline-ApplicationSet-spec in mind, which is the reason why you need this "servegit" thingy. Personally I would've probably pointed the `repoUrl` to an *actual* git-repo which contains `*.appset.yaml` files that I would then feed to `argocd appset generate`.

Of course, your approach is just as valid. It's just a bit different then I would've done it.

**Reply by marxus** [reply]
Hey, the reason I took the inline-spec `servegit` approch with a fake 1min-interval refreshing git repo is to mimic the work of the appset-controller. (also check the `entrypoint.sh` file for the refresh mechanism. it's that simple!)
This is an important part of the how the original appset-controller works. since it has it own refresh interval - the appset manifest might not change, but the generated output, which depends on the appset "generators" might. 
also, the "real" appset manifest is also inlined. so this mimics it.

also CMP plugins have sucky performance since they moved plugins to sidecars.. . when you have a big repo it clones everything and it take a lot of IO. this "fake" repo only contains a 0kb .gitkeep file, it's a hack ofcourse :-)
i even tried to solved that specific problem but ... you know.. maintainers.  - ref: https://github.com/argoproj/argo-cd/pull/18810
